Core/src/main/java/com/plotsquared/core/queue/BlockArrayCacheScopedQueueCoordinator.java

```diff
@@ -0,0 +1,169 @@
+    private final int scopeMaxZ;
+
+    private int offsetX = 0;
+    private int offsetZ = 0;
+
+    /**
+     * Construct a new instance
+     *
+     * @param blockStates Array of form BlockState[y][x][z]. Must be fully initialised.
+     * @param minY        Minimum applicable y value. Used to account for negative world heights (e.g. -64). Inclusive
+     * @param min         Inclusive location of the minimum point to limit the scope to.
+     * @param max         Exclusive location of the maximum point to limit the scope to.
+     * @since TODO
+     */
+    public BlockArrayCacheScopedQueueCoordinator(@NonNull BlockState[][][] blockStates, int minY, Location min, Location max) {
+        super(null, min, max);
+        this.blockStates = blockStates;
+        this.height = blockStates.length;
+        this.width = blockStates[0].length;
+        this.length = blockStates[0][0].length;
+        this.minY = minY;
+        this.maxY = height + minY; // exclusive
+
+        this.scopeMinX = min.getX() & 15;
+        this.scopeMinZ = min.getZ() & 15;
+        this.scopeMaxX = scopeMinX + width;
+        this.scopeMaxZ = scopeMinZ + length;
+    }
+
+    @Override
+    public boolean setBlock(int x, final int y, int z, final @NonNull BlockState id) {
+        x += offsetX;
+        z += offsetZ;
+        if (x >= scopeMinX && x < scopeMaxX && y >= minY && y < maxY && z >= scopeMinZ && z < scopeMaxZ) {
+            blockStates[y - minY][x - scopeMinX][z - scopeMinZ] = id;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean setBlock(final int x, final int y, final int z, @NonNull final Pattern pattern) {
+        int rx = x + offsetX;
+        int rz = z + offsetZ;
+        if (rx >= scopeMinX && rx < scopeMaxX && y >= minY && y < maxY && rz >= scopeMinZ && rz < scopeMaxZ) {
+            BlockState state = pattern
+                    .applyBlock(super.getMin().getBlockVector3().add(BlockVector3.at(x, y, z)))
+                    .toImmutableState();
+            blockStates[y - minY][rx - scopeMinX][rz - scopeMinZ] = state;
+        }
+        return false;
+    }
+
+    @Override
+    public @NonNull Location getMin() {
+        return super.getMin().add(offsetX - scopeMinX, 0, offsetZ - scopeMinZ);
+    }
+
+    @Override
+    public @NonNull Location getMax() {
+        return getMin().add(15, 0, 15).withY(maxY);
+    }
+
+    @Override
+    public boolean setBlock(int x, int y, int z, final @NonNull BaseBlock id) {
+        x += offsetX;
+        z += offsetZ;
+        if (x >= scopeMinX && x < scopeMaxX && y >= minY && y < maxY && z >= scopeMinZ && z < scopeMaxZ) {
+            blockStates[y - minY][x][z] = id.toImmutableState();
+        }
+        return false;
+    }
+
+    @Override
+    public @Nullable BlockState getBlock(final int x, final int y, final int z) {
+        if (x >= 0 && x < width && y >= minY && y < maxY && z >= 0 && z < length) {
+            return blockStates[y - minY][x][z];
```

Comment:
I guess you need to add the offset for x and z here too, as in setBlock. Also, the bounds check should probaböy moved to an extra method to make it more consostent.

Reply:
The queue isn't used for reading blocks anywhere, but I purposely left the offset out of this; writing blocks is specifically done by the generator that writes 0->15 for x/z and thus needs to be offset, and given the point of the queue is to basically aggregate multiple chunks worth of generation, I think it makes sense for any "get" calls to be normalised to the minimum point, but still able to access the entire cache regardless of the offset